### PR TITLE
refactor: adopt `core/Browser` on `BidiBrowser`

### PR DESCRIPTION
**Pull request by jrandolf-google**

### CODE REVIEW

File: packages/puppeteer-core/src/bidi/Browser.ts [hunk 1]
```diff
@@ -70,47 +72,28 @@ export class BidiBrowser extends Browser {
...
+    await session.subscribe(
+      session.capabilities.browserName.toLocaleLowerCase().includes('firefox')
+        ? BidiBrowser.subscribeModules
+        : [...BidiBrowser.subscribeModules, ...BidiBrowser.subscribeCdpEvents]
+    );
 
+    const browser = new BidiBrowser(session.browser, opts);
     await browser.#getTree();
-
+    browser.#initialize();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/bidi/Browser.ts:89]: I think this should happen before browser.#getTree() I believe

Comment (jrandolf) [reply packages/puppeteer-core/src/bidi/Browser.ts:89]: You're right.

File: test/src/browser.spec.ts [hunk 2]
```diff
@@ -64,7 +64,7 @@ describe('Browser specs', function () {
     });
   });
 
-  describe('Browser.isConnected', () => {
+  describe('Browser.connected', () => {
```
Comment (OrKoN) [inline test/src/browser.spec.ts]: Let's revert unrelated changes. Also https://pptr.dev/api/puppeteer.browser.isconnected there is some issue with the documentation for Browser.connected

Comment (Lightning00Blade) [reply test/src/browser.spec.ts]: FYI: The issue is sort of fixed on `next`. It's related to properties not having pages.

**Review by OrKoN — CHANGES_REQUESTED** [review]
tests seem to be failing for Firefox

File: packages/puppeteer-core/src/bidi/Browser.ts [hunk 3]
```diff
@@ -229,29 +217,33 @@ export class BidiBrowser extends Browser {
...
 
-    // `browser.close` can close connection before the response is received.
-    await this.#connection.send('browser.close', {}).catch(debugError);
-    await this.#closeCallback?.call(null);
-    this.#connection.dispose();
+    try {
+      await this.#browserCore.close();
+      await this.#closeCallback?.call(null);
+    } catch (error) {
+      // Fail silently.
```
Comment (OrKoN) [inline packages/puppeteer-core/src/bidi/Browser.ts:240]: I think we should not fail silently for closeCallback and also we should run it even if #browserCore.close() failed.  Perhaps move it to finally?

File: packages/puppeteer-core/src/bidi/core/Session.ts [hunk 4]
```diff
@@ -29,20 +28,21 @@ export class Session
...
-      await new Promise(resolve => {
-        return setTimeout(resolve, (1 << i) * 100);
-      });
-    }
-    if (!status.ready) {
-      throw new Error(status.message);
-    }
+    // TODO: Once Chrome supports session.status properly, uncomment this block.
+    // let status = {message: '', ready: false};
+    // for (let i = 0; i < MAX_RETRIES; ++i) {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/bidi/core/Session.ts:36]: let's only run this if the browser is not Chrome instead of leaving it commented out?

File: packages/puppeteer-core/src/bidi/Browser.ts (hunk 3 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/bidi/Browser.ts:240]: `closeCallback` calls `close` from looking at the location of the callback, so it's fine to fail silently here. We need to rework that logic at some point.

File: packages/puppeteer-core/src/bidi/core/Session.ts (hunk 4 above)
Comment (jrandolf) [reply packages/puppeteer-core/src/bidi/core/Session.ts:36]: Looks like this logic is placed in `BrowserConnector` already. We would need to remove the logic there and replace it with this one once we support pure bidi.

Comment (jrandolf) [reply packages/puppeteer-core/src/bidi/core/Session.ts:36]: I've updated the comment with this information.

File: packages/puppeteer-core/src/bidi/Browser.ts (hunk 3 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/bidi/Browser.ts:240]: It should force close the browser process if the protocol methods failed. The callback is in ProductLauncher. Agree on refactor, not sure if this change will affect any behavior in practice, probably not.

Comment (OrKoN) [reply packages/puppeteer-core/src/bidi/Browser.ts:240]: The tests seem to take quite longer than usual https://github.com/puppeteer/puppeteer/actions/runs/7583926821/job/20656609557?pr=11702

**Review by OrKoN — APPROVED** [review]

File: packages/puppeteer-core/src/bidi/core/Browser.ts [hunk 5]
```diff
@@ -136,15 +134,28 @@ export class Browser extends EventEmitter<{
...
     return browser.#reason!;
   })
   async close(): Promise<void> {
-    await this.#session.send('browser.close', {});
-    this.#reason = `Browser has already closed.`;
-    this.emit('closed', {reason: this.#reason});
-    this.removeAllListeners();
+    try {
+      await this.#session.send('browser.close', {});
+    } finally {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/bidi/core/Browser.ts:156]: Why is this in finally block? The behavior changed now, previously the browser would not be disposed if the command failed.

File: packages/puppeteer-core/src/bidi/core/Session.ts [hunk 6]
```diff
@@ -156,9 +166,10 @@ export class Session
     return session.#reason!;
   })
   async end(): Promise<void> {
-    await this.send('session.end', {});
-    this.#reason = `Session (${this.id}) has already ended.`;
-    this.emit('ended', {reason: this.#reason});
-    this.removeAllListeners();
+    try {
+      await this.send('session.end', {});
+    } finally {
```
Comment (Lightning00Blade) [inline packages/puppeteer-core/src/bidi/core/Session.ts:171]: Same here.

File: packages/puppeteer-core/src/bidi/core/Browser.ts (hunk 5 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/bidi/core/Browser.ts:156]: This has not changed, see packages/puppeteer-core/src/bidi/Browser.ts::close

File: packages/puppeteer-core/src/bidi/core/Session.ts (hunk 6 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/bidi/core/Session.ts:171]: session.end is not supported in mapper.
